### PR TITLE
FastWARC: command-line tools to index and extract WARC records

### PR DESCRIPTION
**Pull request by sebastian-nagel**
Command-line tools to index and extract WARC records are useful to quickly inspect WARC files. The implemented tools/commands behave almost the same as [warcio index](//github.com/webrecorder/warcio#index) and [warcio extract](//github.com/webrecorder/warcio#extract). One difference is for example that `warcio extract` decodes the record payload, removing HTTP transfer and content encoding. Only local files are supported for now because `seek()` and `tell()` are required (would need some efforts to support URLs as well).

```
$> fastwarc index --help
Usage: fastwarc index [OPTIONS] [INFILES]...

  Index WARC records into CDXJ.

Options:
  -o, --output FILENAME  Output file, default is stdout
  -f, --fields TEXT      comma-separated list of indexed fields, eg. "offset",
                         "length", "filename", "http:status", "http:<http-
                         header>", or "<warc-record-header>"

  -h, --help             Show this message and exit.

$> wget http://commoncrawl.s3.amazonaws.com/crawl-data/CC-NEWS/2021/09/CC-NEWS-20210930113548-00741.warc.gz

$> fastwarc index -fwarc-type,http:status,warc-target-uri,filename,offset,length CC-NEWS-20210930113548-00741.warc.gz | head -3
{"warc-type": "warcinfo", "filename": "CC-NEWS-20210930113548-00741.warc.gz", "offset": "0", "length": "423"}
{"warc-type": "request", "warc-target-uri": "https://www.novinite.com/articles/211500/Accident+with+Stranded+Ship+Vera+Su+Might+Be+Caused+by+Inexperienced+Crew+Member", "filename": "CC-NEWS-20210930113548-00741.warc.gz", "offset": "423", "length": "508"}
{"warc-type": "response", "http:status": 200, "warc-target-uri": "https://www.novinite.com/articles/211500/Accident+with+Stranded+Ship+Vera+Su+Might+Be+Caused+by+Inexperienced+Crew+Member", "filename": "CC-NEWS-20210930113548-00741.warc.gz", "offset": "931", "length": "12680"}
```

```
$> fastwarc extract --help
Usage: fastwarc extract [OPTIONS] INFILE OFFSET

  Extract WARC record by offset.

Options:
  --payload   output only record payload (transfer and/or content encoding are
              preserved

  --headers   output only record (and HTTP) headers
  -h, --help  Show this message and exit.

$> fastwarc extract --headers CC-NEWS-20210930113548-00741.warc.gz 931
WARC/1.0
WARC-Record-ID: <urn:uuid:cefe66f5-6ed7-4e2c-839a-964bbfb9fcf2>
Content-Length: 44827
WARC-Date: 2021-09-30T11:35:47Z
WARC-Type: response
WARC-Target-URI: https://www.novinite.com/articles/211500/Accident+with+Stranded+Ship+Vera+Su+Might+Be+Caused+by+Inexperienced+Crew+Member
Content-Type: application/http; msgtype=response
WARC-Payload-Digest: sha1:ICXB6ZRV3DBDGIQJJUFAFM5HBXN574BH
WARC-Block-Digest: sha1:6NYQONX4OQBGVVGCUJYDQ4S7EBAYWH6R

HTTP/1.1 200 OK
Date: Thu, 30 Sep 2021 11:35:48 GMT
Server: Apache/2.4.10 (Debian)
Set-Cookie: PHPSESSID=kbh1betq8no20aeofgpjsnpj64; path=/
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Pragma: no-cache
Vary: Accept-Encoding
X-Crawler-Content-Encoding: gzip
X-Crawler-Content-Length: 12159
Content-Length: 44350
Keep-Alive: timeout=5, max=100
Connection: Keep-Alive
Content-Type: text/html

$> fastwarc extract --payload CC-NEWS-20210930113548-00741.warc.gz 931 | head -7
<!DOCTYPE html>

<html>
<head>
        <base href="https://www.novinite.com/">
        <META http-equiv="Content-Type" content="text/html; charset=utf-8">
        <title>Accident with Stranded Ship Vera Su Might Be Caused by Inexperienced Crew Member - Novinite.com - Sofia News Agency</title>
```

### CODE REVIEW

File: fastwarc/fastwarc/cli.py
Comment (phoerious) [inline fastwarc/fastwarc/cli.py:196]: `WarcRecord.write()` adds an additional double CRLF at the end: one to terminate the record, one to insert a newline before the next record. I think there should be at least one extra CRLF at the end here as well (primarily to prevent weird output effects if the website itself has no trailing newlines).

For writing a complete record, you can also use `WarcRecord.write(ostream)` directly.

File: fastwarc/fastwarc/cli.py
Comment (phoerious) [inline fastwarc/fastwarc/cli.py:183]: If you already depend on local files, you could also use a `stream_io.FileStream` object, which is a little faster than Python file objects. For seeking, you could "abuse" `stream_io._io_stream_py_test_seek()`. Although I don't have an issue with doing it like this.

File: fastwarc/fastwarc/cli.py
Comment (phoerious) [inline fastwarc/fastwarc/cli.py:226]: Why not insert it directly into `idx` from within matching branches? You already do it for `'http:status'`

File: fastwarc/fastwarc/cli.py
Comment (phoerious) [inline fastwarc/fastwarc/cli.py:240]: No newline before.

File: fastwarc/fastwarc/cli.py
Comment (phoerious) [inline fastwarc/fastwarc/cli.py:250]: This is confusing. I would prefer a `length` property on `WarcRecord` that returns the entire record length.

File: fastwarc/fastwarc/cli.py
Comment (phoerious) [inline fastwarc/fastwarc/cli.py:186]: Instead of hard-coding `sys.stdout` here, why not add an output file option here as well, similar to who `fastwarc index` works?

File: fastwarc/fastwarc/cli.py
Comment (phoerious) [inline fastwarc/fastwarc/cli.py:237]: Uppercase first letters (see above) and `click.STRING` should simply be `str`.
You may also want to set `show_default=True` here.

File: fastwarc/fastwarc/cli.py
Comment (phoerious) [inline fastwarc/fastwarc/cli.py:216]: You can use `record.is_http` for checking this.

File: fastwarc/fastwarc/cli.py
Comment (phoerious) [inline fastwarc/fastwarc/cli.py:222]: If you let the `ArchiveIterator` parse HTTP, you need to add the size of the header block to `record.content_length`, use the WARC `Content-Length` header directly, or save the original value and run `record.parse_http()` manually afterwards.

**Review by phoerious — CHANGES_REQUESTED** [review]
Great additions, thanks!

I added a bunch of comments, most are nitpicks.

File: fastwarc/fastwarc/cli.py
Comment (sebastian-nagel) [reply fastwarc/fastwarc/cli.py:196]: Ok. There should be no byte extra in case you want to save binary payload to a file - PDF, image, etc. If the entire record is extracted it should be a valid WARC record, so that users can do tricks such as `for offset in 1234 2345; do fastwarc extract big.warc.gz $offset | gzip; done >small.warc.gz`. I need to double-check whether these use case work.

File: fastwarc/fastwarc/cli.py
Comment (sebastian-nagel) [reply fastwarc/fastwarc/cli.py:183]: For now I think I'll leave it as is.

File: fastwarc/fastwarc/cli.py
Comment (sebastian-nagel) [reply fastwarc/fastwarc/cli.py:226]: Agreed.

File: fastwarc/fastwarc/cli.py
Comment (sebastian-nagel) [reply fastwarc/fastwarc/cli.py]: I'd rather remove this. I added it to understand what record.content_length is. Of course, could be useful, to bind all/more WarcRecord properties to reserved keys.

File: fastwarc/fastwarc/cli.py
Comment (sebastian-nagel) [reply fastwarc/fastwarc/cli.py]: Ok. But I'd had to catch all kinds of exceptions.

File: fastwarc/fastwarc/cli.py
Comment (phoerious) [reply fastwarc/fastwarc/cli.py:195]: Yeah, that could indeed be an edge case. Not adding newlines, on the other hand, would produce output different from what the original WARC file was. Concatenating multiple outputs of `fastwarc extract` wouldn't result in the exact same byte data as the input, since the newlines between the record are missing (although you'd probably use `fastwarc recompress` for usecases like that).

File: fastwarc/fastwarc/cli.py
Comment (sebastian-nagel) [reply fastwarc/fastwarc/cli.py]: Definitely agree. But not today. Feel free to implement this, I'd have to look how to do this anyway. Can be tricky because the entire remaining record needs to be consumed to get the length in the raw WARC file including gzip trailers and alike.

File: fastwarc/fastwarc/cli.py
Comment (phoerious) [reply fastwarc/fastwarc/cli.py]: `content_length` is the effective remaining length, which is not necessarily what the `Content-Length` header indicates. After parsing HTTP, the HTTP header block is subtracted, so `content_length` again is the remaining stream length. It's also documented this way.

File: fastwarc/fastwarc/cli.py
Comment (phoerious) [reply fastwarc/fastwarc/cli.py:183]: Ok.

File: fastwarc/fastwarc/cli.py
Comment (phoerious) [inline fastwarc/fastwarc/cli.py:199]: Thanks for adding this, though you should be more explicit. `StopIteration` is not something you would want to print (and it doesn't have an error message anyway). Instead, you would simply return an empty output.

Other exceptions you actually might want to print would be `stream_io.StreamError` and (theoretically) `OSError` and `ValueError`, which may be raised by Python stream objects.

File: fastwarc/fastwarc/cli.py
Comment (phoerious) [inline fastwarc/fastwarc/cli.py:200]: Thanks for adding this, though you should be more explicit. StopIteration is not something you would want to print (and it doesn't have an error message anyway). Instead, you would simply return an empty output.

Other exceptions you actually might want to print would be stream_io.StreamError and (theoretically) OSError and ValueError, which may be raised by Python stream objects.

Oh, and no legacy format strings, please. Use f-strings instead. ;-)

File: fastwarc/fastwarc/cli.py
Comment (phoerious) [reply fastwarc/fastwarc/cli.py]: Do you really want the length of the input stream segment? Wouldn't the length of the extracted record itself make more sense?

File: fastwarc/fastwarc/cli.py
Comment (phoerious) [reply fastwarc/fastwarc/cli.py]: Not really. Calling `next()` on any kind of iterator would raise `StopIteration`, which needs to be caught. That applies to iterators created with `iter()` as well:

```python
>>> next(iter([]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
StopIteration
```

File: fastwarc/fastwarc/cli.py
Comment (sebastian-nagel) [reply fastwarc/fastwarc/cli.py]: No. The "length" must be the length of the WARC record in the WARC file. Offset and length are used to fetch single WARC records:
```
curl -s -r931-$((931+12680-1)) http://commoncrawl.s3.amazonaws.com/crawl-data/CC-NEWS/2021/09/CC-NEWS-20210930113548-00741.warc.gz | gzip -dc
```

File: fastwarc/fastwarc/cli.py
Comment (sebastian-nagel) [reply fastwarc/fastwarc/cli.py]: Hmhh. But I see for example if pass an offset exceeding the WARC file size:
 ```
  ...
  File "fastwarc/warc.pyx", line 966, in fastwarc.warc.ArchiveIterator.__next__
  File "fastwarc/warc.pyx", line 950, in __iter__
  File "fastwarc/warc.pyx", line 969, in fastwarc.warc.ArchiveIterator._read_next_record
  File "fastwarc/stream_io.pyx", line 750, in fastwarc.stream_io.BufferedReader.detect_stream_type
fastwarc.stream_io.StreamError: Not a valid WARC stream
```

File: fastwarc/fastwarc/cli.py
Comment (phoerious) [reply fastwarc/fastwarc/cli.py]: Okay, that's not possible without read-ahead. In that case, I'd suggest:

```python
from fastwarc.stream_io import _buf_reader_py_test_reset_limit, _buf_reader_py_test_set_limit

# ...

with open(infile, 'rb') as stream:
    for record in ArchiveIterator(stream):
        record.reader.consume()
        _buf_reader_py_test_reset_limit(record.reader)
        _index_record(output, fields, preserve_multi_header, record, record.reader.tell(), infile)
        _buf_reader_py_test_set_limit(record.reader, 0)
```

File: fastwarc/fastwarc/cli.py
Comment (phoerious) [reply fastwarc/fastwarc/cli.py]: Ah, right, that's because the stream negotiation cannot determine the compression type. With an explicit compressor, you would get the same exception, but a different message ("Not a valid GZip stream"). Didn't think of that, but that's an error you may actually want to catch and print. 

`StopIteration` would occur if you started at a valid offset and then went one item beyond the end or if the stream had a valid GZip header, but no further content. So in this case it's indeed a somewhat unlikely scenario, but that would be different if you allowed reading of more than one record (which should be rather trivial to implement).

File: fastwarc/fastwarc/cli.py
Comment (sebastian-nagel) [reply fastwarc/fastwarc/cli.py]: Done. Yes, it works. But only in the sense that it gives the same offsets than the solution using the last record. Comparing the offsets of `TOOL index -fwarc-type,warc-target-uri,offset,length CC-NEWS-20210930113548-00741.warc.gz` for warcio and fastwarc, points to an issue where sometimes the value of record.stream_pos is not increased. Could you have a look?

```
12381,12382c12381,12382
< {"warc-type": "response", "warc-target-uri": "https://www.themarketsdaily.com/2021/09/30/ishares-sp-500-etf-nysearcaivv-sees-strong-trading-volume.html", "offset": "232757027", "length": "12286"}
< {"warc-type": "request", "warc-target-uri": "https://www.agenzianova.com/campania/6155a1e2e36aa9.17205456/3620963/2021-09-29/napoli-approvati-in-giunta-due-interventi-sulle-scuole", "offset": "232769313", "length": "514"}
---
> {"warc-type": "response", "warc-target-uri": "https://www.themarketsdaily.com/2021/09/30/ishares-sp-500-etf-nysearcaivv-sees-strong-trading-volume.html", "offset": "232757027", "length": "0"}
> {"warc-type": "request", "warc-target-uri": "https://www.agenzianova.com/campania/6155a1e2e36aa9.17205456/3620963/2021-09-29/napoli-approvati-in-giunta-due-interventi-sulle-scuole", "offset": "232757027", "length": "12800"}
```

Comment (phoerious) [reply fastwarc/fastwarc/cli.py]: Hmm. That probably happens when a record is larger than a single GZip member. That needs some more investigating, but isn't something that can be fixed in this PR. FastWARC reads ahead in the stream in fixed block sizes or until the end of a member is found. The BufferedReader on top of that doesn't know anything about GZip members, so there is quite some magic going on to calculate raw stream offsets from the reader offsets. Something must go wrong with in the offset calculation:

https://github.com/chatnoir-eu/chatnoir-resiliparse/blob/develop/fastwarc/fastwarc/stream_io.pyx#L435

https://github.com/chatnoir-eu/chatnoir-resiliparse/blob/develop/fastwarc/fastwarc/stream_io.pyx#L924

**Comment by phoerious** [comment]
@sebastian-nagel I have a few fixup changes, could you enable the "allow push access for maintainers" checkbox please so I can push them to this PR?

File: fastwarc/fastwarc/cli.py
Comment (phoerious) [reply fastwarc/fastwarc/cli.py]: Found the reason:

55800cd382c5ef9ccae5e97526a3b700db7ef1fd

Please check if everything works now.

The problem was that I didn't update the stream position if the internal buffer emptied between records while consuming newlines. I only did that for uncompressed streams. Unfortunately, I cannot build test cases for this, since it's a very low-chance event that depends on the sizes of all previous records. Simply extracting the affected records into a small test WARC therefore won't work and I obviously cannot add the full WARC to the repo. I already have tests that check that each record has a higher offset than the previous one, but they didn't trigger on the test WARCs. If you happen to have a WARC where this occurs within the first few kB, please add a small sample of it to `tests/data` and submit a PR.

Comment (phoerious) [reply fastwarc/fastwarc/cli.py]: NVM, I have tests now. Imanaged to create two synthetic WARCs that reproduce the issue by containing records that are exactly the internal buffer size: 4ef411983d647e33c3f9372ac0908c8ea990d45b

**Comment by sebastian-nagel** [comment]
> @sebastian-nagel I have a few fixup changes, could you enable the "allow push access for maintainers" checkbox please so I can push them to this PR?

Hi @phoerious, thanks! There isn't such a checkbox, I have to look why. Maybe this has to be enabled also in the forked repo and/or the commoncrawl account settings. I'll try it after the weekend. But feel free to fix the wrong offsets separately. It isn't directly related, was just a coincident that I've run a diff over the outputs to verify that the change removing the last_record works.

**Comment by phoerious** [comment]
Should be at the bottom of the sidebar.

image

**Comment by phoerious** [comment]
Thanks, I can push my changes now (although that should generally be possible on a per-branch basis for PRs without giving me access to the whole repo).

If you don't have anything else, I'll merge it once CI is done.

**Review by phoerious — APPROVED** [review]

**Comment by sebastian-nagel** [comment]
Yes, please step forward and merge! Thanks!

**Comment by phoerious** [comment]
New PyPi release is up:
- https://pypi.org/project/FastWARC/
- https://pypi.org/project/Resiliparse

**Comment by sebastian-nagel** [comment]
Thanks, @phoerious! Unfortunately, the index command still outputs few records of length 0. Sorry, should have tested more extensively. I've opened #13 to address this.

**Pull request merged**
